File: Basic Calculator II/Java/BasicCalcII-solution-1.java
import java.util.Stack;

public class Solution {
    public int calculate(String s) {
        int len;
        // Handle cases where string is null or empty
        if( s == null || (len = s.length()) == 0) {
            return 0;
        } 
        // stack to hold all numbers
        Stack<Integer> stack = new Stack<Integer>();
        // The current number
        int num = 0;
        // The current sign
        char sign = '+';
        // Loop over all characters in String s
        for(int i = 0; i < len; i++){
            if( Character.isDigit(s.charAt(i)) ){
                // Build up the number if Num has non zero value multiply by 10 so s='12' becomes 12 rather then 3
                // - '0' ensures that we get the number from the ascii value
                num = num * 10 + s.charAt(i) - '0';
            }
            // if The charter is not a digit, blank or last number
            if(( !Character.isDigit(s.charAt(i)) &&' '!= s.charAt(i)) || i == len - 1){
                // push the negative of the num
                if(sign == '-'){
                    stack.push(-num);
                }
                if(sign == '+'){
                    stack.push(num);
                }
                // evaluate multiplication and division immediately
                if(sign == '*'){
                    stack.push(stack.pop() * num);
                }
                if(sign == '/'){
                    stack.push(stack.pop() / num);
                }
                // update sign and reset num
                sign = s.charAt(i);
                num = 0;
            }
        }
        // sum and return the contents of the stack
        int re = 0;
        for(int i : stack){
            re += i;
        }
        return re;
    }
}

File: Basic Calculator II/JavaScript/BasicCalcII-solution-1.js
/**
 * @param {string} s - eg: "1+4 /6-3*9"
 * @return {number}
 */
var calculate = function (s) {
    let rest = [];
    let allOps = ["+", "-", "*", "/"]
    let i = 0;
    let op = null;
    while (i < s.length) {
        let num = "";
        // Loop over string building a number. Handles multi digit numbers
        while (!allOps.includes(s[i]) && i < s.length) {
            // build the new number ignore if the char is blank space
            num += s[i] === ' ' ? '' : s[i];
            i++;
        }


        // check what the last op was
        if (op === '-') {
            rest.push(parseInt(-num)); // If it was '-' the currnt number should be a negative
        } else if (op === '+') {
            rest.push(parseInt(num))
        } else if (op === '*') { // evaluate muliplacation and division right away
            rest.push(parseInt(rest.pop()) * parseInt(num));
        } else if (op === '/') {
            let num1 = rest.pop();
            // This is done because JavaScript handles floor differently for negatives vs positive numbers
            // Track if the first num is negative and assign it to the result of the operation
            let modifier = num1 < 0 ? Math.abs(num1) : num1;
            let sign = num1 < 0 ? -Math.floor(parseInt(modifier) / num) : Math.floor(parseInt(modifier) / num);
            rest.push(sign)
        } else {
            // if there is not op yet just push the number
            rest.push(parseInt(num));
        }
        // Assign the current Op and increment the loop
        op = s[i]; 
        i++;
    }
    // return the sum of the Array
    return rest.reduce((acc, cur) => {
        return acc + cur;
    })
}